File: errite/tools/mis.py
"""

    DeviantCord 2 Discord Bot
    Copyright (C) 2020  Errite Games LLC/ ErriteEpticRikez

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.


"""
from errite.da.jsonTools import findDuplicateElementArray
from markdownify import markdownify as md

def enumerateAllID(data):
    length = len(data)
    if length < 10:
        use_offset = 0
    else:
        use_offset = length - 10
    gathered_id = []

    for entry in data:
        gathered_id.append(entry)
    return gathered_id
def checkHybridResources(da_data, artdata):
    data_resources = {}
    data_resources["all-hybrids"] = []
    data_resources["all-hybrid-urls"] = []
    data_resources["all-hybrid-img-urls"] = []
    data_resources["seen-hybrid-urls"] = []
    data_resources["seen-hybrid-img-urls"] = []
    data_resources["seen-hybrids"] = []
    data_resources["new-hybrids"] = 0
    for entry in da_data["results"]:
        #This blocks writings from being detected
        try:
            check_var = entry["excerpt"]
        except KeyError:
            if not findDuplicateElementArray(artdata, entry["deviationid"]):
                ++data_resources["new-hybrids"]
            else:
                data_resources["seen-hybrids"].append(entry["deviationid"])
                data_resources["seen-hybrid-urls"].append(entry["url"])
                data_resources["seen-hybrid-img-urls"].append(entry["content"]["src"])
            data_resources["all-hybrids"].append(entry["deviationid"])
            data_resources["all-hybrid-urls"].append(entry["url"])
            data_resources["all-hybrid-img-urls"].append(entry["content"]["src"])
    return data_resources

def gatherJournal(data):
    data_resources = {}
    data_resources["ids"] = []
    data_resources["urls"] = []
    data_resources["titles"] = []
    data_resources["img-urls"] = []
    data_resources["excerpts"] = []
    data_resources["profilepic"] = "none"
    if not len(data["results"]) == 0:
        data_resources["profilepic"] = data["results"][0]["author"]["usericon"]
    for entry in data["results"]:
        data_resources["titles"].append(entry["title"])
        data_resources["excerpts"].append(md(entry["excerpt"]))
        data_resources["urls"].append(entry["url"])
        data_resources["ids"].append(entry["deviationid"])
        if not len(entry["thumbs"]) == 0:
            data_resources["img-urls"].append(entry["thumbs"][0]["src"])
        else:
            data_resources["img-urls"].append(None)
    return data_resources

def createIDList(data):
    hybrid_ids = []
    for entry in data["results"]:
        try:
            check_var = entry["excerpt"]
        except KeyError:
            hybrid_ids.append(entry["deviationid"])
    return hybrid_ids

def createIDURLList(data):
    data_resources = {}
    data_resources["ids"] = []
    data_resources["urls"] = []
    data_resources["img-urls"] = []
    for entry in data["results"]:
        try:
            check_var = entry["excerpt"]
        except KeyError:
            data_resources["ids"].append(entry["deviationid"])
            data_resources["urls"].append(entry["url"])
            data_resources["img-urls"].append(entry["content"]["src"])
    return data_resources

def gatherGalleryFolderResources(data):
    data_resources = {}
    data_resources["deviation-ids"] = []
    data_resources["deviation-urls"] = []
    data_resources["img-urls"] = []
    for entry in data["results"]:
        try:
            check_var = entry["excerpt"]
        except KeyError:
            data_resources["deviation-ids"].append(entry["deviationid"])
            data_resources["deviation-urls"].append(entry["url"])
            try:
                data_resources["img-urls"].append(entry["content"]["src"])
            except KeyError:
                print("Trying other formats")
                try:
                    data_resources["img-urls"].append(entry["flash"]["src"] + " DEVIANTCORDENDINGUSENONPREVIEW")
                except KeyError:
                    try:
                        data_resources["img-urls"].append(str(entry["videos"][0]["src"]) + str("DEVIANTCORDENDINGUSENONPREVIEW"))
                    except KeyError:
                        try:
                            data_resources["img-urls"].append(entry["thumbs"]["src"])

                        except:
                            data_resources["img-urls"].append("IGNORETHISDEVIATION")

    return data_resources

def createJournalInfoList(data):
    """
                Method ran to compile needed journal data for the database into a dictionary the specified artist using deviantart's API.
                :param data: The tag that should be searched for.
                :type data: dict
                :return: dict
        """
    data_resources = {}
    data_resources["titles"] = []
    data_resources["profilepic"] = "none"
    data_resources["deviation-ids"] = []
    data_resources["thumbnails-img-urls"] = []
    #Thumbnail ID's contains the deviation-id's that have thumbnails.
    data_resources["thumbnail-ids"] = []
    data_resources["journal-urls"] = []
    data_resources["excerpts"] = []
    if not len(data) == 0:
        data_resources["profilepic"] = data[0]["author"]["usericon"]
    for entry in data:
        data_resources["deviation-ids"].append(entry["deviationid"])
        data_resources["journal-urls"].append(entry["url"])
        data_resources["titles"].append(entry["title"])
        if not len(entry["thumbs"]) == 0:
            try:
                data_resources["thumbnails-img-urls"].append(entry["thumbs"][0]["src"])
                data_resources["thumbnail-ids"].append(entry["deviationid"])
            except KeyError as KE:
                print("Did not detect thumbnail for journal id " + entry["deviationid"])
                data_resources["thumbnails-img-urls"].append("none")
                #data_resources["thumbnail-ids"].append("none")
        elif len(entry["thumbs"]) == 0:
            data_resources["thumbnails-img-urls"].append("none")
        data_resources["excerpts"].append(entry["excerpt"])
    return data_resources



def convertBoolString(bool):
    if bool == True:
        return "true";
    if bool == False:
        return "false";
    else:
        return "invalid";

def convertStringBool(string):
    if string.lower() == "true":
        return True;
    if string.lower() == "false":
        return False;
    else:
        return False;

def fileExists(file):
    try:
        fh = open(file, 'r')
        return True;
        # Store configuration file values
    except FileNotFoundError:
        return False;

File: src/main/java/org/errite/deviantcord/util/mis.java
/*
 * DeviantCord 4
 * Copyright (C) 2020-2024  Errite Softworks LLC/ ErriteEpticRikez
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.errite.deviantcord.util;

import io.restassured.response.Response;
import org.errite.deviantcord.types.JournalObject;
import org.errite.deviantcord.types.StatusObject;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

import java.sql.ResultSet;
import java.util.*;

public class mis {


    public static StatusObject gatherStatuses(Response data){
        StatusObject obtStatuses = new StatusObject();
        ArrayList<LinkedHashMap> node = data.jsonPath().getJsonObject("results");
        Iterator<LinkedHashMap> iterator = node.iterator();
        int currentIndex = 0;
        while(iterator.hasNext()){
            LinkedHashMap entry = iterator.next();
            if(currentIndex == 0){
                LinkedHashMap author = (LinkedHashMap) entry.get("author");
                String userIcon = (String)author.get("usericon");
                if(!(Objects.isNull(userIcon)))
                    obtStatuses.setProfilePic(userIcon);
                else
                    obtStatuses.setProfilePic("none");
            }
            boolean foundThumb = false;
            obtStatuses.appendStatusId((String) entry.get("statusid"));
            obtStatuses.appendStatusUrls((String) entry.get("url"));
            obtStatuses.appendExcerpts((String) entry.get("body"));
            ArrayList items = (ArrayList) entry.get("items") ;
            //TODO put back type variable, if needed
            //LinkedHashMap check_var = (LinkedHashMap) entry.get("type");
            if(!(items.size() == 0)) {
                System.out.println("This is a breakpoint used to check what is going on");
                /*
                LinkedHashMap preview = (LinkedHashMap) items.get("preview");
                LinkedHashMap deviation = (LinkedHashMap) items.get("deviation");
                obtStatuses.appendThumbnailIds((String) deviation.get("deviationid"));
                obtStatuses.appendThumbnailImgUrls((String) preview.get("src"));

                 */
            }
            else{
                obtStatuses.appendThumbnailIds("none");
                obtStatuses.appendThumbnailImgUrls("none");
            }
            ++currentIndex;
        }
        return obtStatuses;
    }
    public static JournalObject gatherJournals(Response data){
        JournalObject obtJournals = new JournalObject();
        ArrayList<LinkedHashMap> node = data.jsonPath().getJsonObject("results");
        Iterator<LinkedHashMap> iterator = node.iterator();
        int currentIndex = 0;
        while(iterator.hasNext()){
            LinkedHashMap entry = iterator.next();
            if(currentIndex == 0)
            {
                obtJournals.setLatestTitle((String) entry.get("title"));
                LinkedHashMap author = (LinkedHashMap) entry.get("author");
                String userIcon = (String)author.get("usericon");
                if(!(Objects.isNull(userIcon)))
                    obtJournals.setProfilePicture(userIcon);
                else
                    obtJournals.setProfilePicture("none");
            }
            
            boolean foundThumb = false;
            obtJournals.appendExcerpts((String)entry.get("excerpt"));
            obtJournals.appendDeviationId((String)entry.get("deviationid"));
            obtJournals.appendJournalUrls((String)entry.get("url"));
            obtJournals.appendTitle((String) entry.get("title"));
            //TODO TEST HERE
            ArrayList<LinkedHashMap> check_var = (ArrayList<LinkedHashMap>) entry.get("thumbs");
            //TODO CHANGE THIS to check for SIZE
            if(!(check_var.size() == 0))
            {
                LinkedHashMap cont = check_var.get(0);
                UUID uuid = UUID.randomUUID();
                obtJournals.appendThumbnailIds(uuid.toString());
                obtJournals.appendThumbnailImgUrls((String)cont.get("src"));

            }
            else{
                obtJournals.appendThumbnailIds("none");
                obtJournals.appendThumbnailImgUrls("none");
            }
            currentIndex++;
        }
        return obtJournals;
    }
    public static HashMap<String, ArrayList<String>> gatherGalleryFolderResources(Response data){
        HashMap<String, ArrayList<String>> data_resources = new HashMap<>();
        data_resources.put("deviation-ids", new ArrayList<String>());
        data_resources.put("deviation-urls", new ArrayList<String>());
        data_resources.put("img-urls", new ArrayList<String>());
        data.jsonPath().getJsonObject("results");
        ArrayList<LinkedHashMap> msg =  data.jsonPath().getJsonObject("results");
        Iterator<LinkedHashMap> iterator = msg.iterator();
        while(iterator.hasNext())
        {
            LinkedHashMap entry = iterator.next();
            String check_var = (String) entry.get("excerpt");
            ArrayList<String> ids = data_resources.get("deviation-ids");
            ArrayList<String> dev_urls = data_resources.get("deviation-urls");
            dev_urls.add((String)entry.get("url"));
            ids.add((String)entry.get("deviationid"));

            if(Objects.isNull(check_var)) {
                    LinkedHashMap cont = (LinkedHashMap) entry.get("content");
                    String src = (String) cont.get("src");
                    ArrayList<String> urls = data_resources.get("img-urls");
                    urls.add(src);
                    data_resources.put("img-urls", urls);
                 if(Objects.isNull(cont)) {
                    cont = (LinkedHashMap) entry.get("flash");
                    src = (String) cont.get("src");
                    urls = data_resources.get("img-urls");
                    urls.add(src + "DEVIANTCORDENDINGUSENONPREVIEW");
                    data_resources.put("img-urls", urls);
                     if (Objects.isNull(cont)) {
                        LinkedHashMap videos = (LinkedHashMap) entry.get("videos");
                        src = (String) videos.get("src");
                        urls = data_resources.get("img-urls");
                        urls.add(src + "DEVIANTCORDENDINGUSENONPREVIEW");
                        data_resources.put("img-urls", urls);
                        if (Objects.isNull(videos)) {
                            cont = (LinkedHashMap) entry.get("thumbs");
                            src = (String) cont.get("src");
                            urls = data_resources.get("img-urls");
                            urls.add(src);
                            data_resources.put("img-urls", urls);
                             if(Objects.isNull(cont)) {
                                urls = data_resources.get("img-urls");
                                urls.add("IGNORETHISDEVIATION");
                                data_resources.put("img-urls", urls);
                            }

                        }

                    }

                }
            }


        }
        return data_resources;
    }

    public static boolean InvertBoolean(boolean given_boolean){
        if(given_boolean == true)
            return false;
        else
            return true;
    }
    public static boolean StringToBoolean(String input)
    {
        if(input.toUpperCase().equals("TRUE"))
            return true;
        else if (input.toUpperCase().equals("FALSE"))
            return false;
        else
            return false;
    }
}
